resteasy-client-jetty/src/main/java/org/jboss/resteasy/client/jaxrs/engines/jetty/JettyClientEngine.java

```diff
@@ -112,112 +97,75 @@ public <T> CompletableFuture<T> submit(ClientInvocation request, boolean buffere
         if (request.getAttributes().get(FOLLOW_REDIRECTS) == Boolean.FALSE) {
             request.followRedirects(false);
         }
 
         if (invocation.getEntity() != null) {
-            final ByteArrayOutputStream os = new ByteArrayOutputStream();
-            try {
-                invocation.writeRequestBody(os);
-            } catch (IOException e) {
-                future.completeExceptionally(e);
-                if (callback != null) {
-                    callback.failed(e);
+            final OutputStreamRequestContent contentOut = new OutputStreamRequestContent(
+                    invocation.getHeaders().getMediaType().toString());
+            asyncExecutor.execute(() -> {
+                try {
+                    invocation.writeRequestBody(contentOut.getOutputStream());
+                } catch (Exception e) {
+                    future.completeExceptionally(e);
+                    if (callback != null) {
+                        callback.failed(e);
+                    }
+                } finally {
+                    contentOut.close();
                 }
-                return future;
-            }
-            request.content(new BytesContentProvider(os.toByteArray()));
+            });
+            request.body(contentOut);
         }
 
-        request.send(new Response.Listener.Adapter() {
+        request.send(new InputStreamResponseListener() {
             private ClientResponse cr;
-            private JettyResponseStream stream = new JettyResponseStream();
 
             @Override
+            @SuppressWarnings("unchecked")
             public void onHeaders(Response response) {
-                cr = new JettyClientResponse(invocation.getClientConfiguration(), stream, () -> {
-                    if (!completing.get()) {
-                        future.cancel(true);
-                    }
-                });
+                super.onHeaders(response);
+                InputStream inputStream = getInputStream();
+                cr = new JettyClientResponse(invocation.getClientConfiguration(), inputStream);
                 cr.setProperties(invocation.getMutableProperties());
                 cr.setStatus(response.getStatus());
                 cr.setHeaders(extract(response.getHeaders()));
-                if (!buffered) {
-                    complete();
-                }
-            }
-
-            @Override
-            public void onContent(Response response, ByteBuffer buf) {
-                final ByteBufferPool bufs = client.getByteBufferPool();
-                final ByteBuffer copy = bufs.acquire(buf.remaining(), false);
-                copy.limit(buf.remaining());
-                copy.put(buf);
-                copy.flip();
-                stream.offer(copy, new ReleaseCallback(bufs, copy));
-            }
-
-            @Override
-            public void onSuccess(Response response) {
-                if (buffered) {
+                asyncExecutor.submit(() -> {
```

Comment:
Should this be done on the `onHeader()`? It seems odd to me. `onComplete()` or `onSuccess()` seems to make more sense to me. However, I'm not a Jetty expert.

Reply:
I think neither complete nor success handlers would work - both of these callbacks are fired only _after_ the response body is processed. (It's in the javadoc.)

We might be able to defer this work to an onContent callback, but since that will be called many times, it requires a bit more work to keep track of state.

Do note that onHeader*s* is called after all headers are received and parsed - it is not called for each header - so the very next thing coming is body content.

Reply:
This seems okay for now. What's not clear is if there could be any potential race issues with `onHeader()` if the content itself is not yet available. However, so far that doesn't seem to be an issue so I might be overthinking it.

Reply:
All it is doing is dispatching the deserializer work to a thread here. It will attempt to read from the content input stream, which the InputStreamResponseListener takes care of synchronizing. So the first read will block until content is available.